src/main/java/de/tum/in/www1/artemis/service/ExamService.java

```diff
@@ -182,10 +187,147 @@ public void delete(Long examId) {
         return exams.stream().filter(Exam::isVisibleToStudents).collect(Collectors.toSet());
     }
 
+    /**
+     * Returns the relevant result of a student participation
+     *
+     * @param studentParticipation studentParticipation to get relevant result for
+     * @return optional of relevant result
+     */
+    private Optional<Result> getRelevantResult(StudentParticipation studentParticipation) {
+        // no participant -> no relevant result
+        if (studentParticipation.getParticipant() == null) {
+            return Optional.empty();
+        }
+
+        return studentParticipation.getResults().stream().filter(Result::isRated).filter(result -> result.getCompletionDate() != null).filter(result -> result.getScore() != null)
+                .sorted((r1, r2) -> r2.getCompletionDate().compareTo(r1.getCompletionDate())).findFirst();
+
+    }
+
+    /**
+     * Rounds a double to a specified number of decimal places
+     *
+     * @param value  the number to round
+     * @param places the number of decimal places to round to
+     * @return rounded number
+     */
+    public static double round(double value, int places) {
+        if (places < 0)
```

Comment:
please add { }

Reply:
Changed in 5011dbdf2b222a17c0674381032e8a6712c329a8